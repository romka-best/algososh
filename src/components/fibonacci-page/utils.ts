export function* getFibonacciNumbers(number: number): Generator<Array<number>> {
    const numbers: Array<number> = [];
    for (let i = 0; i <= number; i++) {
        if (i < 2) {
            numbers.push(1);
        } else {
            numbers.push(numbers[i - 2] + numbers[i - 1]);
        }
        yield numbers;
    }

    return numbers;
}
