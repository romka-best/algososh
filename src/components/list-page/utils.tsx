import React from "react";

import {Circle} from "../ui/circle/circle";

import LinkedList from "./LinkedList";

import {HEAD, TAIL} from "../../constants/element-captions";
import {ADD_OPERATION_TYPE, DELETE_OPERATION_TYPE} from "./constants";

import {ElementStates} from "../../types/element-states";
import {OperationTypes, Step} from "./list-page.types";

export function addHead<T>(value: T, list: LinkedList<T>): Step<T>[] {
    const steps: Step<T>[] = [];

    steps.push({
        index: 0,
        value,
        list: list.toArray(),
    });

    list.prepend(value);

    steps.push({
        index: 0,
        list: list.toArray(),
    });

    steps.push({
        list: list.toArray(),
    });

    return steps;
}

export function addTail<T>(value: T, list: LinkedList<T>): Step<T>[] {
    const steps: Step<T>[] = [];

    let currentList = list.toArray();
    steps.push({
        index: currentList.length - 1,
        value,
        list: currentList,
    });

    list.append(value);

    currentList = list.toArray();
    steps.push({
        index: currentList.length - 1,
        list: currentList,
    });

    steps.push({
        list: currentList,
    });

    return steps;
}

export function addByIndex<T>(value: T, index: number, list: LinkedList<T>): Step<T>[] {
    const steps: Step<T>[] = [];
    const currentList = list.toArray();

    let currentIndex = 0;
    while (currentIndex !== index) {
        steps.push({
            index: currentIndex,
            value,
            list: currentList,
        });
        currentIndex++;
    }

    list.addByIndex(index, value);

    steps.push({
        index: currentIndex,
        list: list.toArray(),
    });

    steps.push({
        list: list.toArray(),
    });

    return steps;
}

export function deleteHead<T>(list: LinkedList<T>): Step<T>[] {
    const steps: Step<T>[] = [];

    const currentList = list.toArray();
    const value = list.head?._value;
    currentList[0]._value = null;
    steps.push({
        index: 0,
        value,
        list: currentList,
    });

    list.deleteHead();

    steps.push({
        index: 0,
        list: list.toArray(),
    });

    steps.push({
        list: list.toArray(),
    });

    return steps;
}

export function deleteTail<T>(list: LinkedList<T>): Step<T>[] {
    const steps: Step<T>[] = [];

    let currentList = list.toArray();
    const value = list.tail?._value;
    currentList[currentList.length - 1]._value = null;
    steps.push({
        index: currentList.length - 1,
        value,
        list: currentList,
    });

    list.deleteTail();

    currentList = list.toArray();

    steps.push({
        index: currentList.length - 1,
        list: currentList,
    });

    steps.push({
        list: currentList,
    });

    return steps;
}

export function deleteByIndex<T>(index: number, list: LinkedList<T>): Step<T>[] {
    const steps: Step<T>[] = [];

    let currentList = list.toArray();

    let currentIndex = 0;
    while (currentIndex !== index) {
        steps.push({
            index: currentIndex,
            list: currentList,
        });
        currentIndex++;
    }

    const prevList = new LinkedList(list.toInitialArray()).toArray();
    const value = prevList[currentIndex]._value;
    prevList[currentIndex]._value = null;
    steps.push({
        index: currentIndex,
        value,
        list: prevList,
    });

    list.deleteByIndex(index);

    currentList = list.toArray();

    steps.push({
        index: currentIndex,
        list: currentList,
    });

    steps.push({
        list: currentList,
    });

    return steps;
}

export function getLetterState<T>(
    index: number,
    currentStep: Step<T>,
    operationType: OperationTypes | null,
): ElementStates {
    if (!operationType || currentStep.index === undefined) {
        return ElementStates.Default;
    }

    if ((currentStep.value && ((operationType === "addHead" || operationType === "addByIndex") &&
            index < currentStep.index) || (operationType === "addTail" && index > currentStep.index)) ||
        (operationType === "deleteByIndex" && index <= currentStep.index)
    ) {
        return ElementStates.Changing;
    }

    if (!currentStep.value && currentStep.index === index) {
        return ElementStates.Modified;
    }

    return ElementStates.Default;
}

export function getLetterElementHead(step: Step<string>, index: number, operation: OperationTypes | null): string | React.ReactElement {
    if (operation &&
        step.value &&
        step.index === index &&
        ADD_OPERATION_TYPE.includes(operation)) {
        return (
            <Circle
                letter={step.value}
                state={ElementStates.Changing}
                isSmall
            />
        )
    }

    if (!index) {
        return HEAD;
    }

    return "";
}

export function getLetterElementTail(step: Step<string>,
                                     index: number,
                                     operation: OperationTypes | null,
                                     length: number): string | React.ReactElement {
    if (operation &&
        step.value &&
        step.index === index &&
        DELETE_OPERATION_TYPE.includes(operation)) {
        return (
            <Circle
                letter={step.value}
                state={ElementStates.Changing}
                isSmall
            />
        )
    }

    if (index === length - 1) {
        return TAIL;
    }

    return "";
}
