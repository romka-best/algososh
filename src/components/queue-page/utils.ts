import Queue from './Queue';

import { ElementStates } from '../../types/element-states';
import { Step, OperationTypes } from './queue-page.types';

export function enqueue<T>(value: T, queue: Queue<T>): Step<T>[] {
    const steps: Step<T>[] = [];

    steps.push({
        index: queue.tail,
        value,
        queue,
    });

    queue.enqueue(value);

    steps.push({
        queue,
    });

    return steps;
}

export function dequeue<T>(queue: Queue<T>): Step<T>[] {
    const steps: Step<T>[] = [];

    const prevQueue = new Queue<T>(queue._maxN);
    prevQueue.head = queue.head;
    prevQueue.tail = queue.tail;
    prevQueue._size = queue.size();
    prevQueue.queue = Array.from(queue.queue);

    steps.push({
        index: prevQueue.head,
        queue: prevQueue,
    });

    queue.dequeue();

    steps.push({
        queue,
    });

    return steps;
}

export function clear<T>(queue: Queue<T>) {
    const steps: Step<T>[] = [];

    steps.push({
        queue,
    });

    queue.clear();

    steps.push({
        queue,
    });

    return steps;
}

export function getLetterState<T>(
    index: number,
    currentStep: Step<T>,
    operationType: OperationTypes | null,
): ElementStates {
    if (!operationType || currentStep.index === undefined) {
        return ElementStates.Default;
    }

    if (index === currentStep.index) {
        return ElementStates.Changing;
    }

    return ElementStates.Default;
}
