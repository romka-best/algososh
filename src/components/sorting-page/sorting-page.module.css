.root {

}

.form {
    display: flex;
    justify-content: center;
    align-items: center;

    margin-bottom: 48px;
}

.radioButtons {
    display: flex;
    gap: 40px
}

.choiceButtons {
    display: flex;
    gap: 12px;

    margin: 0 80px 0 52px;
}

.generateButton {

}

.columns {
    display: flex;
    justify-content: center;
    align-items: flex-end;
    gap: 10px;

    height: 340px;
}
