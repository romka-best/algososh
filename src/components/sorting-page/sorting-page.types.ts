import {ElementStates} from "../../types/element-states";

export interface IValue {
    number: number;
    type: ElementStates;
}

export type byTypeSort = "По возрастанию" | "По убыванию" | "Нет";
export type choiceTypeSort = "Выбор" | "Пузырёк";
