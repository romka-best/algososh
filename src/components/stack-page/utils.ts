import Stack from './Stack';

import { ElementStates } from '../../types/element-states';

import { Step } from './stack-page.types';
import { OperationTypes } from './stack-page.types';

export function push<T>(value: T, list: Stack<T>) {
    const steps: Step<T>[] = [];

    steps.push({
        index: list.size(),
        value,
        list: list,
    });

    list.push(value);

    steps.push({
        list: list,
    });

    return steps;
}

export function pop<T>(list: Stack<T>) {
    const steps: Step<T>[] = [];

    const prevStack = new Stack<T>();
    prevStack.storage = Array.from(list.storage);

    steps.push({
        index: prevStack.size() - 1,
        list: prevStack,
    });

    list.pop();

    steps.push({
        list: list,
    });

    return steps;
}

export function clear<T>(list: Stack<T>) {
    const steps: Step<T>[] = [];

    steps.push({
        list,
    });

    list.clear();

    steps.push({
        list,
    });

    return steps;
}

export function getLetterState<T>(
    index: number,
    currentStep: Step<T>,
    operationType: OperationTypes | null,
): ElementStates {
    if (!operationType || currentStep.index === undefined) {
        return ElementStates.Default;
    }

    if (index === currentStep.index) {
        return ElementStates.Changing;
    }

    return ElementStates.Default;
}
