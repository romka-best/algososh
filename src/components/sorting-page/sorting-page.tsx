import React from 'react';

import { SolutionLayout } from '../ui/solution-layout/solution-layout';
import { Button } from '../ui/button/button';
import { RadioInput } from '../ui/radio-input/radio-input';
import { Column } from '../ui/column/column';

import { DELAY_IN_MS } from '../../constants/delays';
import { ElementStates } from '../../types/element-states';
import { Direction } from '../../types/direction';

import { byTypeSort, choiceTypeSort, IValue } from './sorting-page.types';
import { bubbleSortAlgorithm, choiceSortAlgorithm, randomArr } from './utils';
import styles from './sorting-page.module.css';

export const SortingPage: React.FC = () => {
    const [ currentChoiceTypeSort, setCurrentChoiceTypeSort ] = React.useState<choiceTypeSort>('Выбор');
    const [ isLoading, setIsLoading ] = React.useState<boolean>(false);
    const [ byTypeSort, setByTypeSort ] = React.useState<byTypeSort>('Нет');

    const [ arrayOfNumbers, setArrayOfNumbers ] = React.useState<Array<IValue>>([]);

    const handleSubmitButton = React.useCallback((event: React.SyntheticEvent) => {
        event.preventDefault();

        if (isLoading) {
            return;
        }

        const defaultValues: Array<IValue> = [];
        for (let i = 0; i < arrayOfNumbers.length; i++) {
            defaultValues.push({
                number: arrayOfNumbers[i].number, type: ElementStates.Default
            });
        }
        setArrayOfNumbers(defaultValues);
        setIsLoading(true);
    }, [ arrayOfNumbers, isLoading ]);

    const handleResetButton = React.useCallback((event: React.SyntheticEvent) => {
        event.preventDefault();

        setArrayOfNumbers(randomArr());
    }, []);

    React.useEffect(() => {
        if (isLoading) {
            const generatorAlgorithm = currentChoiceTypeSort === 'Выбор' ?
                choiceSortAlgorithm(arrayOfNumbers, byTypeSort) :
                bubbleSortAlgorithm(arrayOfNumbers, byTypeSort);

            const interval = setInterval(() => {
                const generatorValue = generatorAlgorithm.next();

                setArrayOfNumbers(() => {
                    const newState: Array<IValue> = [];
                    for (let i = 0; i < generatorValue.value.length; i++) {
                        newState.push(generatorValue.value[i]);
                    }
                    return newState;
                });

                if (generatorValue.done) {
                    clearInterval(interval);
                    setIsLoading(false);
                }
            }, DELAY_IN_MS);
        }
    }, [ isLoading ]);

    React.useEffect(() => {
        setArrayOfNumbers(randomArr());
    }, []);

    return (
        <SolutionLayout title="Сортировка массива">
            <div className={ styles.root }>
                <form className={ styles.form } onSubmit={ handleSubmitButton } onReset={ handleResetButton }>
                    <div className={ styles.radioButtons }>
                        <RadioInput
                            id="ChoiceRadioInput"
                            label="Выбор"
                            checked={ currentChoiceTypeSort === 'Выбор' }
                            disabled={ isLoading }
                            onChange={ () => {
                                setCurrentChoiceTypeSort('Выбор');
                            } }
                        />
                        <RadioInput
                            id="BubbleRadioInput"
                            label="Пузырёк"
                            checked={ currentChoiceTypeSort === 'Пузырёк' }
                            disabled={ isLoading }
                            onChange={ () => {
                                setCurrentChoiceTypeSort('Пузырёк');
                            } }
                        />
                    </div>
                    <div className={ styles.choiceButtons }>
                        <Button
                            text="По возрастанию"
                            sorting={ Direction.Ascending }
                            disabled={ isLoading }
                            isLoader={ isLoading && byTypeSort === 'По возрастанию' }
                            onClick={ () => {
                                setByTypeSort('По возрастанию');
                            } }
                            type="submit"
                        />
                        <Button
                            text="По убыванию"
                            sorting={ Direction.Descending }
                            disabled={ isLoading }
                            isLoader={ isLoading && byTypeSort === 'По убыванию' }
                            onClick={ () => {
                                setByTypeSort('По убыванию');
                            } }
                            type="submit"
                        />
                    </div>
                    <Button
                        extraClass={ styles.generateButton }
                        text="Новый массив"
                        disabled={ isLoading }
                        type="reset"
                    />
                </form>
                <div className={ styles.columns }>
                    {
                        arrayOfNumbers.map((value, index) => (
                                <Column
                                    key={ index }
                                    value={ value.number }
                                    state={ value.type }
                                />
                            )
                        )
                    }
                </div>
            </div>
        </SolutionLayout>
    );
};
