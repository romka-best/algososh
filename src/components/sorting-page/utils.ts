import { ElementStates } from '../../types/element-states';

import { byTypeSort, IValue } from './sorting-page.types';

export function randomArr(): Array<IValue> {
    const MAX_LEN = 17;
    const MIN_LEN = 3;
    const MIN_NUMBER = 0;
    const MAX_NUMBER = 100;
    const lenNumbers = Math.floor(Math.random() * (MAX_LEN - MIN_LEN + 1) + MIN_LEN);

    const numbers: Array<IValue> = [];
    for (let i = 0; i < lenNumbers; i++) {
        const number = Math.floor(Math.random() * (MAX_NUMBER - MIN_NUMBER + 1) + MIN_NUMBER);
        numbers.push({
            number,
            type: ElementStates.Default,
        });
    }

    return numbers;
}

export function* choiceSortAlgorithm(array: Array<IValue>, type: byTypeSort): Generator<Array<IValue>> {
    const length = array.length;

    for (let i = 0; i < length; i++) {
        let minOrMax = i;
        for (let j = i; j < length; j++) {
            if ((type === 'По возрастанию' && array[j].number < array[minOrMax].number) ||
                (type === 'По убыванию' && array[j].number > array[minOrMax].number)) {
                minOrMax = j;
            }
        }
        if (minOrMax !== i) {
            array[i] = {
                ...array[i],
                type: ElementStates.Changing,
            };
            array[minOrMax] = {
                ...array[minOrMax],
                type: ElementStates.Changing,
            };
            yield array;

            const temp = array[i];
            array[i] = {
                number: array[minOrMax].number,
                type: ElementStates.Modified,
            };
            array[minOrMax] = {
                number: temp.number,
                type: ElementStates.Default,
            };
            yield array;
        } else {
            array[i] = {
                ...array[i],
                type: ElementStates.Changing,
            };
            yield array;

            array[i] = {
                ...array[i],
                type: ElementStates.Modified,
            };
            array[minOrMax] = {
                ...array[minOrMax],
                type: ElementStates.Modified,
            };
            yield array;
        }
    }
    return array;
}

export function* bubbleSortAlgorithm(array: Array<IValue>, type: byTypeSort): Generator<Array<IValue>> {
    for (let j = array.length - 1; j >= 0; j--) {
        for (let i = 0; i < j; i++) {
            if ((type === 'По возрастанию' && array[i].number > array[i + 1].number) ||
                (type === 'По убыванию' && array[i].number < array[i + 1].number)) {
                const temp = array[i];
                array[i] = {
                    number: array[i + 1].number,
                    type: ElementStates.Changing,
                };
                array[i + 1] = {
                    number: temp.number,
                    type: ElementStates.Changing,
                };
                yield array;
            }
            array[i] = {
                ...array[i],
                type: ElementStates.Default,
            };
            array[i + 1] = {
                ...array[i + 1],
                type: ElementStates.Default,
            };
        }
        array[j] = {
            ...array[j],
            type: ElementStates.Modified,
        };
        yield array;
    }
    return array;
}
