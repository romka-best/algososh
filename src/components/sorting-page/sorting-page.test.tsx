import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';

import { getValueFromGenerator } from '../../utils/getValueFromGenerator';
import { ElementStates } from '../../types/element-states';

import { SortingPage } from './sorting-page';
import { bubbleSortAlgorithm, choiceSortAlgorithm } from './utils';

describe('SortingPage', () => {
    describe('Отображение', () => {
        it('Начальное состяние', () => {
            const page = renderer
                .create(
                    <BrowserRouter>
                        <SortingPage/>
                    </BrowserRouter>
                )
                .toJSON();

            expect(page).toMatchSnapshot();
        });
    });

    describe('Логика', () => {
        describe('Работа с алгоритмом', () => {
            describe('Сортировка выбором', () => {
                it('Пустой массив: По возрастанию', () => {
                    const generator = choiceSortAlgorithm([], 'По возрастанию');

                    const value = getValueFromGenerator(generator);
                    expect(value).toStrictEqual([]);
                });

                it('Пустой массив: По убыванию', () => {
                    const generator = choiceSortAlgorithm([], 'По убыванию');

                    const value = getValueFromGenerator(generator);
                    expect(value).toStrictEqual([]);
                });

                it('Массив из одного элемента: По возрастанию', () => {
                    const generator = choiceSortAlgorithm([
                        { number: 10, type: ElementStates.Default },
                    ], 'По возрастанию');

                    const value = getValueFromGenerator(generator);
                    expect(value).toStrictEqual([
                        { number: 10, type: ElementStates.Modified },
                    ]);
                });

                it('Массив из одного элемента: По убыванию', () => {
                    const generator = choiceSortAlgorithm([
                        { number: 10, type: ElementStates.Default },
                    ], 'По убыванию');

                    const value = getValueFromGenerator(generator);
                    expect(value).toStrictEqual([
                        { number: 10, type: ElementStates.Modified },
                    ]);
                });

                it('Массив из нескольких элементов: По возрастанию', () => {
                    const generator = choiceSortAlgorithm([
                        { number: 15, type: ElementStates.Default },
                        { number: 5, type: ElementStates.Default },
                        { number: 10, type: ElementStates.Default },
                        { number: 13, type: ElementStates.Default },
                        { number: 7, type: ElementStates.Default },
                    ], 'По возрастанию');

                    const value = getValueFromGenerator(generator);
                    expect(value).toStrictEqual([
                        { number: 5, type: ElementStates.Modified },
                        { number: 7, type: ElementStates.Modified },
                        { number: 10, type: ElementStates.Modified },
                        { number: 13, type: ElementStates.Modified },
                        { number: 15, type: ElementStates.Modified },
                    ]);
                });

                it('Массив из нескольких элементов: По убыванию', () => {
                    const generator = choiceSortAlgorithm([
                        { number: 15, type: ElementStates.Default },
                        { number: 5, type: ElementStates.Default },
                        { number: 10, type: ElementStates.Default },
                        { number: 13, type: ElementStates.Default },
                        { number: 7, type: ElementStates.Default },
                    ], 'По убыванию');

                    const value = getValueFromGenerator(generator);
                    expect(value).toStrictEqual([
                        { number: 15, type: ElementStates.Modified },
                        { number: 13, type: ElementStates.Modified },
                        { number: 10, type: ElementStates.Modified },
                        { number: 7, type: ElementStates.Modified },
                        { number: 5, type: ElementStates.Modified },
                    ]);
                });
            });

            describe('Сортировка пузыриком', () => {
                it('Пустой массив: По возрастанию', () => {
                    const generator = bubbleSortAlgorithm([], 'По возрастанию');

                    const value = getValueFromGenerator(generator);
                    expect(value).toStrictEqual([]);
                });

                it('Пустой массив: По убыванию', () => {
                    const generator = bubbleSortAlgorithm([], 'По убыванию');

                    const value = getValueFromGenerator(generator);
                    expect(value).toStrictEqual([]);
                });

                it('Массив из одного элемента: По возрастанию', () => {
                    const generator = bubbleSortAlgorithm([
                        { number: 10, type: ElementStates.Default },
                    ], 'По возрастанию');

                    const value = getValueFromGenerator(generator);
                    expect(value).toStrictEqual([
                        { number: 10, type: ElementStates.Modified },
                    ]);
                });

                it('Массив из одного элемента: По убыванию', () => {
                    const generator = bubbleSortAlgorithm([
                        { number: 10, type: ElementStates.Default },
                    ], 'По убыванию');

                    const value = getValueFromGenerator(generator);
                    expect(value).toStrictEqual([
                        { number: 10, type: ElementStates.Modified },
                    ]);
                });

                it('Массив из нескольких элементов: По возрастанию', () => {
                    const generator = bubbleSortAlgorithm([
                        { number: 15, type: ElementStates.Default },
                        { number: 5, type: ElementStates.Default },
                        { number: 10, type: ElementStates.Default },
                        { number: 13, type: ElementStates.Default },
                        { number: 7, type: ElementStates.Default },
                    ], 'По возрастанию');

                    const value = getValueFromGenerator(generator);
                    expect(value).toStrictEqual([
                        { number: 5, type: ElementStates.Modified },
                        { number: 7, type: ElementStates.Modified },
                        { number: 10, type: ElementStates.Modified },
                        { number: 13, type: ElementStates.Modified },
                        { number: 15, type: ElementStates.Modified },
                    ]);
                });

                it('Массив из нескольких элементов: По убыванию', () => {
                    const generator = bubbleSortAlgorithm([
                        { number: 15, type: ElementStates.Default },
                        { number: 5, type: ElementStates.Default },
                        { number: 10, type: ElementStates.Default },
                        { number: 13, type: ElementStates.Default },
                        { number: 7, type: ElementStates.Default },
                    ], 'По убыванию');

                    const value = getValueFromGenerator(generator);
                    expect(value).toStrictEqual([
                        { number: 15, type: ElementStates.Modified },
                        { number: 13, type: ElementStates.Modified },
                        { number: 10, type: ElementStates.Modified },
                        { number: 7, type: ElementStates.Modified },
                        { number: 5, type: ElementStates.Modified },
                    ]);
                });
            });
        });
    });
});
