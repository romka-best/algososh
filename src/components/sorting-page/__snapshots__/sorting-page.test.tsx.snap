// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortingPage Отображение Начальное состяние 1`] = `
<main
  className="content"
>
  <div
    className="titleBox"
  >
    <h1
      className="text text_type_h2 text_color_h1 title"
    >
      МБОУ АЛГОСОШ
    </h1>
    <span
      className="text text_type_fibonacci text_color_secondary subtitle"
    >
      им. Фибоначчи
    </span>
  </div>
  <div
    className="contentCard"
  >
    <a
      className="link"
      href="/"
      onClick={[Function]}
    >
      <button
        className="button returnButton"
        type="button"
      >
        <svg
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M12.7073 4.29289C13.0978 4.68342 13.0978 5.31658 12.7073 5.70711L7.41436 11H19.0002C19.5524 11 20.0002 11.4477 20.0002 12C20.0002 12.5523 19.5524 13 19.0002 13H7.41436L12.7073 18.2929C13.0978 18.6834 13.0978 19.3166 12.7073 19.7071C12.3167 20.0976 11.6836 20.0976 11.293 19.7071L3.58594 12L11.293 4.29289C11.6836 3.90237 12.3167 3.90237 12.7073 4.29289Z"
            fill="#0032FF"
            fillRule="evenodd"
          />
        </svg>
        <p
          className="text text_type_button text_color_link ml-4"
        >
          К оглавлению
        </p>
      </button>
    </a>
    <h3
      className="text text_type_h3 text_color_h3 cardTitle"
    >
      Сортировка массива
    </h3>
    <div
      className="root"
    >
      <form
        className="form"
        onReset={[Function]}
        onSubmit={[Function]}
      >
        <div
          className="radioButtons"
        >
          <div
            className="content"
          >
            <input
              checked={true}
              className="input"
              disabled={false}
              id="ChoiceRadioInput"
              onChange={[Function]}
              type="radio"
            />
            <label
              className="text text_type_button label"
              htmlFor="ChoiceRadioInput"
            >
              Выбор
            </label>
          </div>
          <div
            className="content"
          >
            <input
              checked={false}
              className="input"
              disabled={false}
              id="BubbleRadioInput"
              onChange={[Function]}
              type="radio"
            />
            <label
              className="text text_type_button label"
              htmlFor="BubbleRadioInput"
            >
              Пузырёк
            </label>
          </div>
        </div>
        <div
          className="choiceButtons"
        >
          <button
            className="text text_type_button text_color_primary button"
            disabled={false}
            onClick={[Function]}
            type="submit"
          >
            <svg
              fill="inherit"
              height="12"
              viewBox="0 0 20 12"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1 0C0.447715 0 0 0.447715 0 1C0 1.55228 0.447715 2 1 2H11C11.5523 2 12 1.55228 12 1C12 0.447715 11.5523 0 11 0H1ZM0 6C0 5.44772 0.447715 5 1 5H15C15.5523 5 16 5.44772 16 6C16 6.55228 15.5523 7 15 7H1C0.447715 7 0 6.55228 0 6ZM0 11C0 10.4477 0.447715 10 1 10H19C19.5523 10 20 10.4477 20 11C20 11.5523 19.5523 12 19 12H1C0.447716 12 0 11.5523 0 11Z"
                fill="inherit"
                fillOpacity="0.85"
              />
            </svg>
            <p
              className="text ml-5"
            >
              По возрастанию
            </p>
          </button>
          <button
            className="text text_type_button text_color_primary button"
            disabled={false}
            onClick={[Function]}
            type="submit"
          >
            <svg
              fill="inherit"
              height="12"
              viewBox="0 0 20 12"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clipRule="evenodd"
                d="M0 1C0 0.44772 0.44772 0 1 0H19C19.5523 0 20 0.44772 20 1C20 1.55228 19.5523 2 19 2H1C0.44772 2 0 1.55228 0 1ZM0 6C0 5.4477 0.44772 5 1 5H15C15.5523 5 16 5.4477 16 6C16 6.5523 15.5523 7 15 7H1C0.44772 7 0 6.5523 0 6ZM1 10C0.44772 10 0 10.4477 0 11C0 11.5523 0.44772 12 1 12H11C11.5523 12 12 11.5523 12 11C12 10.4477 11.5523 10 11 10H1Z"
                fill="inherit"
                fillOpacity="0.85"
                fillRule="evenodd"
              />
            </svg>
            <p
              className="text ml-5"
            >
              По убыванию
            </p>
          </button>
        </div>
        <button
          className="text text_type_button text_color_primary button generateButton"
          disabled={false}
          type="reset"
        >
          <p
            className="text"
          >
            Новый массив
          </p>
        </button>
      </form>
      <div
        className="columns"
      />
    </div>
  </div>
  <p
    className="text text_type_column text_color_input mt-14 copyright"
  >
    © Сделано в Практикуме Романом Даниловым.
  </p>
</main>
`;
